File: Miscellaneous/Python/121. Best Time to Buy and Sell Stock.py
class Solution:
    def maxProfit(self, prices):
        maxDiff, initial = 0, prices[0]

        for i in range(1, len(prices)):
            profit = prices[i] - initial
            maxDiff = max(maxDiff, profit)
            initial = min(initial, prices[i])

        return maxDiff

obj = Solution()
prices = [7,1,5,3,6,4]
ans = obj.maxProfit(prices)
print(ans)

File: Miscellaneous/Golang/121. Best Time to Buy and Sell Stock.go
package main

import "fmt"

func main() {

	arr := []int{7, 1, 5, 3, 6, 4}
	result := maxProfit(arr)
	fmt.Println(result)
}

func maxProfit(arr []int) int {
	maxDiff, idx := 0, arr[0]

	for i := 1; i < len(arr); i++ {
		profit := arr[i] - idx
		maxDiff = max(maxDiff, profit)
		if idx > arr[i] {
			idx = arr[i]
		}
	}

	return maxDiff
}

func max(a, b int) int {
	if a >= b {
		return a
	}
	return b
}
